### ISSUE TITLE
Enhancement of ZonedDatetime for SQL

### ISSUE THREAD

**Issue by cbornet** [jhipster/generator-jhipster#12262]
Currently our support of ZonedDatetime in SQL is minimal : we store the instant represented by this ZD and assume UTC timezone.
We have something a little better on Cassandra where a ZoneDatetimeCodec will store the date and timezone in distinct columns and so the timezone is preserved.
I propose to do the same for SQL. I did a PoC with the following code to add a column in DB to store the zone id.
```java
    @Transient
    private ZonedDateTime myDate;

    @JsonIgnore
    @Column(name = "my_date")
    private LocalDateTime myDateLocalDateTime;

    @JsonIgnore
    @Column(name = "my_date_zoneid")
    private ZoneId myDateZoneId;

    @SuppressWarnings("unused")
    private void setMyDateLocalDateTime(LocalDateTime myDateLocalDateTime) {
        this.myDateLocalDateTime = myDateLocalDateTime;
        this.myDate = null;
    }

    @SuppressWarnings("unused")
    private void setMyDateZoneId(ZoneId myDateZoneId) {
        this.myDateZoneId = myDateZoneId;
        this.myDate = null;
    }

    @JsonFormat(with = JsonFormat.Feature.WRITE_DATES_WITH_ZONE_ID)
    public ZonedDateTime getMyDate() {
        if (myDate == null && myDateLocalDateTime != null && myDateZoneId != null) {
            myDate = ZonedDateTime.of(myDateLocalDateTime, myDateZoneId);
        }
        return myDate;
    }

    public void setMyDate(ZonedDateTime myDate) {
        this.myDate = myDate;
        this.myDateLocalDateTime = myDate.toLocalDateTime();
        this.myDateZoneId = myDate.getZone();
    }
```
> NB : the behavior when using the REST API depends on Jackson's ADJUST_DATES_TO_CONTEXT_TIME_ZONE setting.

WDYT ?

**Comment by cbornet** [comment]
Better : I found [this little gem of a library](https://github.com/marschall/threeten-jpa/tree/master/threeten-jpa-zoned-hibernate) which makes it as simple as
```java
    @Type(type = ZonedDateTimeType.NAME)
    @Columns(columns = {
        @Column(name = "my_date"),
        @Column(name = "my_date_zoneid")
    })
    @JsonFormat(with = JsonFormat.Feature.WRITE_DATES_WITH_ZONE_ID)
    private ZonedDateTime myDate;
```
without the need to add new fields

**Comment by pascalgrimaud** [comment]
you're the boss in this domain @cbornet : I'm trusting you here :)

**Comment by avdev4j** [comment]
It's look great, a bit weird to think that no "out of the box" solution exists for every DB but well...
Just a question: could we just keep the "myDate" setter? I wonder if the others are useful (setMyDateZoneId and myDateLocalDateTime).

**Comment by cbornet** [comment]
> Just a question: could we just keep the "myDate" setter?

Yes that setter will still be here. I just didn't show it for simplicity.

 I wonder if the others are useful (setMyDateZoneId and myDateLocalDateTime).

With the first solution they were needed to reset the ZonedDatetime. But with the second solution the LocalDateTime. and ZoneId fields don't exist so the corresponding getters/setters won't be present either.

**Comment by mraible** [comment]
@cbornet Any updates on this issue? Can we close it?

**Comment by cbornet** [comment]
I've got something almost ready. Just need some time to finish it.

**Comment by funder7** [comment]
Hi, I'm making some considerations about the timezone topic and I've found this issue open, so instead of opening a new one I'd like to ask you...does make sense to set liquibase like this: `<column name="myDateTime" type="timestamp with time zone"/>` ?
This allows to store the time with timezone. The fact is that I'm lazy and I don't want to add a new column for timezone for each datetime field :-) 
I see many people talking about that, it makes me feel like I'm missing something in my considerations about this topic.

**Comment by cbornet** [comment]
No. The liquibase type `timestamp with time zone` corresponds to a time offset (+01:00, +02:00, etc...), not a zone id ("PDT, CET, etc...) which takes DST in consideration.
If you're only interested in the time offset, you can use an `OffsetDateTime` instead.

**Comment by funder7** [comment]
In fact I'm only interested into having time saved correctly, I'm using `Instant` class

**Comment by DanielFran** [comment]
@cbornet Are you still planning to implement it?

**Comment by mshima** [comment]
Hibernate 6 has different storage options for zoned times https://thorben-janssen.com/hibernate-6-offsetdatetime-and-zoneddatetime/.
I don't think we should serialize timestamps as json, maybe add optional support to `TimeZoneStorageType.COLUMN` (adjust liquibase generator to add the timezone column).

**Comment by dennismzia** [comment]
this project looks stale is it still alive

**Comment by mraible** [comment]
Yes, it's very alive!

**Comment by cbornet** [comment]
Sorry I forgot about this issue. I’ll have a look again.
